packages/utils/src/Option.ts

```diff
@@ -0,0 +1,318 @@
+import { assert, AssertionError } from './assert';
+
+/**
+ * We use a symbol so that the Option can represent Some(undefined)
+ */
+const nothing = Symbol('Internal symbol that means no data in Option');
+
+function isSome<T>(data: T | typeof nothing): data is T {
+  return data !== nothing;
+}
+function isNone<T>(data: T | typeof nothing): data is typeof nothing {
+  return data === nothing;
+}
+
+export function Maybe<T>(data: T | undefined | null): Option<T> {
+  return Option.Maybe(data);
+}
+export function Some<T>(data: T): Option<T> {
+  return Option.Some(data);
+}
+export function None<T>(): Option<T> {
+  return Option.None();
+}
+/**
+ * @see {@link Option.flat}
+ * @see {@link Array.flat}
+ */
+// TODO(ritave): Make fully recursive when we move to TS4.8 https://github.com/microsoft/TypeScript/issues/14833#issuecomment-1238924118
+type FlatOption<Opt, Depth extends number> = {
+  done: Opt;
+  recur: Opt extends Option<infer InnerOpt>
+    ? FlatOption<
+        InnerOpt,
+        [
+          -1,
+          0,
+          1,
+          2,
+          3,
+          4,
+          5,
+          6,
+          7,
+          8,
+          9,
+          10,
+          11,
+          12,
+          13,
+          14,
+          15,
+          16,
+          17,
+          18,
+          19,
+          20,
+        ][Depth]
+      >
+    : Opt;
+}[Depth extends -1 ? 'done' : 'recur'];
+
+// Inspired by rust https://doc.rust-lang.org/std/option/
+export class Option<T> {
```

Comment:
I don't think this is used right now. We should also discuss whether we want to use this, as people who haven't used Rust might not be familiar with this pattern. If we do want this, there's a lot of existing libraries out there which we could use, rather than inventing the wheel again. Here's a few I found after a quick Google search:

- https://github.com/OliverBrotchie/optionals
- https://github.com/franeklubi/rustic
- https://github.com/traverse1984/oxide.ts
- https://github.com/exoticknight/rust-option
- And many more.